src/test/kotlin/no/nav/helsearbeidsgiver/arbeidsgivernotifikasjon/MockArbeidsgiverNotifikasjonKlient.kt

```diff
@@ -20,20 +16,5 @@ fun mockArbeidsgiverNotifikasjonKlient(content: String): ArbeidsgiverNotifikasjo
             headers = headersOf(HttpHeaders.ContentType, ContentType.Application.Json.toString())
         )
     }
-
-    val mockHttpClient = HttpClient(mockEngine)
-
-    return mockFn(::createHttpClient) {
-        every { createHttpClient() } returns mockHttpClient
-        ArbeidsgiverNotifikasjonKlient("https://url") { "fake token" }
-    }
-}
-
-private fun <T> mockFn(fn: KFunction<*>, block: () -> T): T {
-    mockkStatic(fn)
-    return try {
-        block()
-    } finally {
-        unmockkStatic(fn)
-    }
+    return ArbeidsgiverNotifikasjonKlient("https://url", HttpClient(mockEngine)) { "fake token" }
```

Comment:
Hvilke problemer opplever du med denne måten å teste på?

Reply:
Jeg får problemer ved å mocke svarene til klienten. Den slår ikke til i testene. Utvider derfor klienten slik at man teste på foretrukken måte i ktor.

Reply:
Det er ikke galt i dette repo å teste det slik. Det er helt ok. Men det er når man skal mocke denne klienten i andre applikasjoner det blir problemer.

Reply:
Jeg sliter med å forstå hvorfor man vil mocke den interne http-klienten andre steder enn i dette repoet. Hvis man skal teste noe som bruker arbeidsgiver-notifikasjon-klient så mocker man vel bare den klienten, og ikke den interne http-klienten? 

Mocker man den interne http-klienten kan testene også brekke dersom intern logikk i arbeidsgiver-notifikasjon-klient endrer seg.

Reply:
Jeg får ikke lov å mocke svaret til metodene til ArbeidsgiverNotifikasjonKlient i Simba. Derfor vil jeg utvide med mulighet for å sende inn en egen klient med egen response for å kunne lage test i Simba. All funksjonalitet blir beholdt. 

Denne måten å teste på vil jeg bruke:
https://ktor.io/docs/http-client-testing.html#test-client

Reply:
Får du ikke lov tli å gjøre
```kotlin
val mock = mockk<ArbeidsgiverNotifikasjonKlient>()
coEvery { mock.nyStatusSak(any(), any()) } returns "mockId"
```
?

Reply:
Jo, men det virker ikke! Det er problemet. Den kjører klient koden alikevel. Hva er problemet med å tillate denne utvidelsen? Dette forhindrer meg i gå fremover.

Reply:
Å kunne sende inn en http-klient i konstruktøren åpner for feil, som vi har opplevd tidligere. Jeg synes også at den nye måten å teste på hørtes skjørt ut. Når du nevner at klienten kjører koden uansett, så virker det for meg som at det er mockk som er problemet. Da må vi "fikse" mockk.

Jeg skrev noe enkel kode for å teste, og jeg får reprodusert problemet. Det ser ut som det stammer av at metodene til arbeidsgiver-notifikasjon-klient er definert som extension-funksjoner, som må mockes annerledes. Hvis du wrapper testkoden din i `mockkStatic(ArbeidsgiverNotifikasjonKlient::<funksjonSomMockes>) { ... }` så burde det løse problemet.

Reply:
Dette åpner opp for muligheten til å velge selv. Dersom man av en eller annen grunn har et behov for å bruke en annen HttpClient som må man få lov til det. Du MÅ ikke bruke en annen HttpClient. 

Jeg ønsker å bruke anbefalt måte å teste klientene til Ktor på. Det er ikke skjørt - snarere tvert imot. Det er forutsigbart fordi man kan velge responsen også er det lett å utvikle tester mot. Det står dokumentert på Ktor sidene med flere eksempler.

Forslagene dine går imot standard måten å løse testing på i Ktor og krever workarounds.

Reply:
> Dersom man av en eller annen grunn har et behov for å bruke en annen HttpClient som må man få lov til det.
Hvorfor må man få lov til det? Jeg ser ingen fordeler. Ulempen er at man får muligheten til å sende inn http-klienter som koden ikke er laget for.

> Jeg ønsker å bruke anbefalt måte å teste klientene til Ktor på.
Vi bruker anbefalt måte å teste på i dette repoet. Utenfor dette repoet så ser jeg ingen grunn til å teste http-klienten. Vi tester ikke intern funksjonalitet fra annen kode vi henter inn som pakker.

> Forslagene dine går imot standard måten å løse testing på i Ktor og krever workarounds.
Det går ikke imot standarden. Se svaret over for det. Når det gjelder workarounds så er det koden som den er skrevet nå som krever en annen type mocking dersom man bruker mockk. Det er ikke en workaround, men sånn mockk fungerer. Dersom vi vil gjøre testingen enklere med mockk så er det enkelt nok å fikse  ved å gå bort ifra extension-funksjoner, og det hadde nok vært lurt å gjøre det.